Update IDeveloperPageExceptionFilter doc comment

yecril71pl opened the pull request:
# Make `DeveloperExceptionPageMiddleware` a code reference

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

##  Make `DeveloperExceptionPageMiddleware` a code reference

so that readers can follow it in the [documentation page](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.diagnostics.ideveloperpageexceptionfilter).

I think opening an issue for this change would be excessive.

yecril71pl committed:
<see cref="DeveloperExceptionPageMiddleware"/>

martincostello commented on src/Middleware/Diagnostics.Abstractions/src/IDeveloperPageExceptionFilter.cs:
```diff
@@ -9,7 +9,7 @@
 
 /// <summary>
-/// Provides an extensiblity point for changing the behavior of the DeveloperExceptionPageMiddleware.
+/// Provides an extensiblity point for changing the behavior of the <see cref="DeveloperExceptionPageMiddleware"/>.
```
```suggestion
/// Provides an extensibility point for changing the behavior of the <see cref="Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware"/>.
```

yecril71pl committed:
typo fix

Co-authored-by: Martin Costello <martin@martincostello.com>

yecril71pl committed:
Include="Microsoft.AspNetCore.Diagnostics"

Trying to fix an unknown reference error

yecril71pl committed:
Use href because of a MSB4006

This package is referenced by the main package and it cannot refer to a class defined in there.

yecril71pl committed:
rv Include=Microsoft.AspNetCore.Diagnostics

halter73 commented on src/Middleware/Diagnostics.Abstractions/src/IDeveloperPageExceptionFilter.cs:
```diff
@@ -9,7 +9,7 @@
 
 /// <summary>
-/// Provides an extensiblity point for changing the behavior of the DeveloperExceptionPageMiddleware.
+/// Provides an extensibility point for changing the behavior of the <see href="Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware"/>.
```
```suggestion
/// Provides an extensibility point for changing the behavior of the <see cref="DeveloperExceptionPageMiddleware"/>.
```

halter73 committed:
Update src/Middleware/Diagnostics.Abstractions/src/IDeveloperPageExceptionFilter.cs

halter73 commented on src/Middleware/Diagnostics.Abstractions/src/IDeveloperPageExceptionFilter.cs:
```diff
@@ -9,7 +9,7 @@
 
 /// <summary>
-/// Provides an extensiblity point for changing the behavior of the DeveloperExceptionPageMiddleware.
+/// Provides an extensibility point for changing the behavior of the <see cref="DeveloperExceptionPageMiddleware"/>.
```
```suggestion
/// Provides an extensibility point for changing the behavior of the DeveloperExceptionPageMiddleware.
```

The issue is that IDeveloperPageExceptionFilter is defined in Microsoft.AspNetCore.Diagnostics.Abstractions instead of Microsoft.AspNetCore.Diagnostics and the reference points the other way. Does anyone know a way to link to unreferenced types?

yecril71pl replied:
Just use `href`.

martincostello replied:
For `href` I think you'd need to link to a valid fully-qualified URI for it to be useful.

yecril71pl replied:
It would be a bug in the doc engine.

martincostello replied:
Tools other than those that power `docs.microsoft.com` also process XML documentation comments, so the markup needs to point to something concrete. `href="DeveloperExceptionPageMiddleware"` by itself doesn't point to anything specific as it's not a fully-qualified URL (like `https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.diagnostics.developerexceptionpagemiddleware).

While the "doc engine" for `docs.microsoft.com` could be smart enough to special-case it, that would only fix that website, and wouldn't be useful for other tooling (such as Visual Studio Intellisense, Visual Studio Code, Rider, etc.) that might render the link to the user in some other context, as it wouldn't go anywhere.

yecril71pl replied:
You would make the same argument to shoot down the concept of a relative URL wholesale.  Also, an improvement to one site is still an improvement.

martincostello replied:
An improvement to one site at the expense of a dead-end link that does nothing in `N > 1` other tools isn't a net-positive improvement though.

yecril71pl replied:
Why would the paths to these 2 documents be different in any tool?  They are logically in the same place.

martincostello replied:
`DeveloperExceptionPageMiddleware` doesn't point anywhere though - if you put that into your URL bar in your browser, you'd get a 404 or the results of a search engine query for that literal string. If you did an HTTP GET request to it with `HttpClient`, you wouldn't get anything.

It's a relative URI, not an absolute URI, and without tooling special-casing to go and find what's _meant_, it won't find anything useful.

martincostello replied:
I think the crux of the issue is what is the URI relative _to_? The file? The namespace? The assembly?

Seems to me that it's contextual and ambiguous (where a `cref` isn't), which is why the `href` doesn't seem correct IMO.

If there's something about how that's treated universally that I'm not aware of that makes it "just work", then fair enough.

yecril71pl replied:
The URI is relative to the base path of the referrer.

martincostello replied:
But what is the referrer? Seems to be something that would be contextual to me.

yecril71pl replied:
The referrer  depends on the user agent.  But since both documents are in the same context, this question is irrelevant.

martincostello replied:
Not really - what if the User Agent is Visual Studio's Intellisense pop-up - what would it open if you clicked on the link?

The types are in different assemblies, they just happen to be in the same namespace. I don't see how this would work in a context like this where I click `IDictionary<TKey, TValue>` and it opens the assembly reference for the `IDictionary<TKey, TValue>` type in the `System.Runtime` assembly. How is the user agent going to know where the type lives _just_ from the information in the link?

![image](https://user-images.githubusercontent.com/1439341/144062891-f7148814-b213-4994-9b53-6f16b405b520.png)

yecril71pl replied:
I was corrected to use the full type rather than the bare type name.  Intellisense should just find the type in the assemblies referred to by the project.  If it is unable to do that for any reason, it should disable relative `href` links, rendering just the text (which is what you would like to happen everywhere).  That would mean the link would be OK in the browser and harmless for Intellisense.

halter73 replied:
I'm going to just apply my suggestion for now and take the spelling fix. Sorry @yecril71pl. While I agree it would be nice if all the tooling worked well with unreferenced types, I do see why it's a difficult problem to solve, and I don't think we can solve it here.

adityamandaleeka commented:
Does anyone know how to make this link work?

cc @Rick-Anderson

Rick-Anderson commented:
I don't work on API ref but I'm pretty sure I've seen @JamesNK  use fully qualified links in API Ref.

Rick-Anderson commented:
See this 
https://github.com/dotnet/AspNetApiDocs/blob/main/aspnet-core/xml/ns-Microsoft.AspNetCore.Mvc.ModelBinding.xml

halter73 commented:
Thanks @Rick-Anderson. This option was discussed in the thread at https://github.com/dotnet/aspnetcore/pull/38659#discussion_r758774350, but I agree with @martincostello. It's one thing to do that in a remark for a namespace, and it's another thing to put it on an interface summary. It would be weird to click on a type name in intellisense and have a browser open. I'm thinking we leave it as is and fix the spelling of extensibility.

halter73 committed:
Update src/Middleware/Diagnostics.Abstractions/src/IDeveloperPageExceptionFilter.cs

Rick-Anderson commented:
>The issue is that IDeveloperPageExceptionFilter is defined in Microsoft.AspNetCore.Diagnostics.Abstractions instead of `Microsoft.AspNetCore.Diagnostics` and the reference points the other way. Does anyone know a way to link to unreferenced types?

@gewarren  do you know how to  do this?

gewarren commented:
It's possible that this would work, but you'd need to suppress [CA1200](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1200):

`<see cref="T:Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware"/>`

yecril71pl committed:
Add a FIXME comment so that nobody retries

halter73 commented:
Thanks @yecril71pl! Sorry we couldn't work out the `cref`.
